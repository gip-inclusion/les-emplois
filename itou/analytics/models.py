import uuid

from django.db import models
from django.utils import timezone

from itou.common_apps.address.departments import DEPARTMENTS
from itou.users.enums import UserKind


class DatumCode(models.TextChoices):
    # Employee record - Base
    EMPLOYEE_RECORD_COUNT = "ER-001", "FS totales"
    EMPLOYEE_RECORD_DELETED = "ER-002", "FS (probablement) supprimées"
    # Employee record - Lifecycle
    EMPLOYEE_RECORD_PROCESSED_AT_FIRST_EXCHANGE = "ER-101", "FS intégrées (0000) au premier retour"
    EMPLOYEE_RECORD_WITH_ERROR_AT_FIRST_EXCHANGE = "ER-102", "FS avec une erreur au premier retour"
    EMPLOYEE_RECORD_WITH_ERROR_3436_AT_FIRST_EXCHANGE = "ER-102-3436", "FS avec une erreur 3436 au premier retour"
    EMPLOYEE_RECORD_WITH_AT_LEAST_ONE_ERROR = "ER-103", "FS ayant eu au moins un retour en erreur"
    # Approval - Base
    APPROVAL_COUNT = "AP-001", "PASS IAE total"
    APPROVAL_CANCELLED = "AP-002", "PASS IAE annulés"
    # Approval - PE notification cycle
    APPROVAL_PE_NOTIFY_SUCCESS = "AP-101", "PASS IAE synchronisés avec succès avec pole emploi"
    APPROVAL_PE_NOTIFY_PENDING = "AP-102", "PASS IAE en attente de synchronisation avec pole emploi"
    APPROVAL_PE_NOTIFY_ERROR = "AP-103", "PASS IAE en erreur de synchronisation avec pole emploi"
    APPROVAL_PE_NOTIFY_READY = "AP-104", "PASS IAE prêts à être synchronisés avec pole emploi"
    # Users
    USER_COUNT = "US-001", "Nombre d'utilisateurs"
    USER_JOB_SEEKER_COUNT = "US-011", "Nombre de demandeurs d'emploi"
    USER_PRESCRIBER_COUNT = "US-012", "Nombre de prescripteurs"
    USER_EMPLOYER_COUNT = "US-013", "Nombre d'employeurs"
    USER_LABOR_INSPECTOR_COUNT = "US-014", "Nombre d'inspecteurs du travail"
    USER_ITOU_STAFF_COUNT = "US-015", "Nombre d'administrateurs"
    # API usage
    API_TOTAL_CALLS = "API-001", "API : total d'appels reçus"
    API_TOTAL_UV = "API-002", "API : total de visiteurs uniques"
    API_TOTAL_CALLS_CANDIDATS = "API-003", "API candidats : total d'appels reçus"
    API_TOTAL_UV_CANDIDATS = "API-004", "API candidats : total de visiteurs uniques"
    API_TOTAL_CALLS_GEIQ = "API-005", "API GEIQ : total d'appels reçus"
    API_TOTAL_UV_GEIQ = "API-006", "API GEIQ : total de visiteurs uniques"
    API_TOTAL_CALLS_ER = "API-007", "API FS : total d'appels reçus"
    API_TOTAL_UV_ER = "API-008", "API FS : total de visiteurs uniques"
    API_TOTAL_CALLS_MARCHE = "API-009", "API Le marché : total d'appels reçus"
    API_TOTAL_CALLS_SIAES = "API-010", "API siaes : total d'appels reçus"
    API_TOTAL_UV_SIAES = "API-011", "API siaes : total de visiteurs uniques (par adresse IP)"
    API_TOTAL_CALLS_STRUCTURES = "API-012", "API structures : total d'appels reçus"


class Datum(models.Model):
    """Store an aggregated `value` of the `code` data point for the specified `bucket`."""

    id = models.UUIDField(primary_key=True, default=uuid.uuid4)

    code = models.TextField(choices=DatumCode.choices)
    bucket = models.TextField()
    value = models.IntegerField()  # Integer offers the best balance between range, storage size, and performance

    measured_at = models.DateTimeField(default=timezone.now)  # Not using auto_now_add=True to allow overrides

    class Meta:
        verbose_name_plural = "data"
        unique_together = ["code", "bucket"]
        indexes = [models.Index(fields=["measured_at", "code"])]


class StatsDashboardVisit(models.Model):
    dashboard_id = models.IntegerField(verbose_name="ID tableau de bord Metabase")
    dashboard_name = models.TextField(verbose_name="nom de la vue du tableau de bord")
    department = models.CharField(verbose_name="département", choices=DEPARTMENTS.items(), max_length=3, null=True)
    region = models.TextField(verbose_name="région", null=True)
    current_company_id = models.IntegerField(verbose_name="ID entreprise courante", null=True)
    current_prescriber_organization_id = models.IntegerField(
        verbose_name="ID organisation prescriptrice courante", null=True
    )
    current_institution_id = models.IntegerField(verbose_name="ID institution courante", null=True)
    user_kind = models.TextField(verbose_name="type d'utilisateur", choices=UserKind.choices)
    user_id = models.IntegerField(verbose_name="ID utilisateur")

    measured_at = models.DateTimeField(default=timezone.now)  # Not using auto_now_add=True to allow overrides

    class Meta:
        verbose_name_plural = "visite de tableau de bord"
