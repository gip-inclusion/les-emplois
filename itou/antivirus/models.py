from django.db import models
from django.utils.text import capfirst

from itou.files.models import File


class Scan(models.Model):
    file = models.OneToOneField(File, on_delete=models.CASCADE)
    clamav_signature = models.TextField()
    clamav_completed_at = models.DateTimeField(null=True, verbose_name="analyse ClamAV le")
    clamav_infected = models.BooleanField(null=True, verbose_name="fichier infect√© selon ClamAV")
    comment = models.TextField(blank=True, verbose_name="commentaire")

    class Meta:
        verbose_name = "analyse antivirus"
        verbose_name_plural = "analyses antivirus"

    def __str__(self):
        text = f"{capfirst(self._meta.verbose_name)} {self.file_id}"
        if self.clamav_infected:
            text = f"[VIRUS] {text}"
        return text
