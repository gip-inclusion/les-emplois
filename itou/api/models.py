import secrets

from django.db import models
from django.utils import timezone

from itou.companies.models import Company


def _generate_key():
    return secrets.token_urlsafe()


class CompanyToken(models.Model):
    key = models.CharField(default=_generate_key, unique=True, editable=False)
    label = models.CharField(verbose_name="m√©mo permettant d'identifier l'usage du jeton", max_length=60, unique=True)
    created_at = models.DateTimeField(default=timezone.now)
    companies = models.ManyToManyField(Company, related_name="api_tokens")

    class Meta:
        verbose_name = "jeton d'API entreprise"
        verbose_name_plural = "jetons d'API entreprise"
