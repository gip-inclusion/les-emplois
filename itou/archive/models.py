import uuid

from django.db import models


class AnonymizedProfessional(models.Model):
    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)

    # from User model
    date_joined = models.DateField(verbose_name="année et mois d'inscription")
    first_login = models.DateField(verbose_name="année et mois de première connexion", blank=True, null=True)
    last_login = models.DateField(verbose_name="année et mois de dernière connexion", blank=True, null=True)
    anonymized_at = models.DateTimeField(auto_now_add=True, verbose_name="anonymisé le")
    department = models.CharField(max_length=3, verbose_name="département", blank=True, null=True)
    title = models.CharField(
        max_length=3,
        verbose_name="civilité",
        blank=True,
        null=True,
    )
    kind = models.CharField(max_length=50, verbose_name="type de professionnel", blank=True, null=True)
    number_of_memberships = models.PositiveIntegerField(verbose_name="nombre d'adhésions", default=0)
    number_of_active_memberships = models.PositiveIntegerField(verbose_name="nombre d'adhésions actives", default=0)
    number_of_memberships_as_administrator = models.PositiveIntegerField(
        verbose_name="nombre d'adhésions en tant qu'administrateur", default=0
    )
    had_memberships_in_authorized_organization = models.BooleanField(
        verbose_name="adhésion à au moins une organisation habilitée", default=False
    )
    identity_provider = models.CharField(max_length=20, verbose_name="fournisseur d'identité (SSO)")

    class Meta:
        verbose_name = "professionnel anonymisé"
        verbose_name_plural = "professionnels anonymisés"
        ordering = ["-anonymized_at"]


class AnonymizedJobSeeker(models.Model):
    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)

    # from User model
    date_joined = models.DateField(verbose_name="année et mois d'inscription")
    first_login = models.DateField(verbose_name="année et mois de première connexion", blank=True, null=True)
    last_login = models.DateField(verbose_name="année et mois de dernière connexion", blank=True, null=True)
    anonymized_at = models.DateTimeField(auto_now_add=True, verbose_name="anonymisé le")
    user_signup_kind = models.CharField(
        max_length=50, verbose_name="créé par un utilisateur de type", blank=True, null=True
    )
    department = models.CharField(max_length=3, verbose_name="département", blank=True, null=True)
    title = models.CharField(
        max_length=3,
        verbose_name="civilité",
        blank=True,
        null=True,
    )
    identity_provider = models.CharField(max_length=20, verbose_name="fournisseur d'identité (SSO)")

    # from JobSeekerProfile model
    had_pole_emploi_id = models.BooleanField(verbose_name="ID Pôle emploi", default=False)
    had_nir = models.BooleanField(verbose_name="NIR", default=False)
    lack_of_nir_reason = models.CharField(verbose_name="raison de l'absence de NIR", blank=True, null=True)
    nir_sex = models.PositiveSmallIntegerField(verbose_name="sexe du NIR", blank=True, null=True)
    nir_year = models.PositiveSmallIntegerField(verbose_name="année du NIR", blank=True, null=True)
    birth_year = models.PositiveSmallIntegerField(verbose_name="année de naissance", blank=True, null=True)
    count_accepted_applications = models.PositiveIntegerField(
        verbose_name="nombre de candidatures acceptées", default=0
    )
    count_IAE_applications = models.PositiveIntegerField(verbose_name="nombre de candidatures dans l'IAE", default=0)
    count_total_applications = models.PositiveIntegerField(verbose_name="nombre de candidatures totales", default=0)

    class Meta:
        verbose_name = "candidat anonymisé"
        verbose_name_plural = "candidats anonymisés"
        ordering = ["-anonymized_at"]

    def __str__(self):
        return f"candidat {self.id} anonymisé le {self.anonymized_at.strftime('%Y-%m-%d %H:%M:%S')}"


class AnonymizedApplication(models.Model):
    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)

    # job_seeker
    job_seeker_birth_year = models.PositiveSmallIntegerField(
        verbose_name="année de naissance du candidat", blank=True, null=True
    )
    job_seeker_department_same_as_company_department = models.BooleanField(
        verbose_name="le candidat a le même département que celui l'entreprise", default=False
    )

    # sender
    sender_kind = models.CharField(verbose_name="type de l'émetteur")
    sender_company_kind = models.CharField(verbose_name="type de l'entreprise émettrice", blank=True, null=True)
    sender_prescriber_organization_kind = models.CharField(
        verbose_name="type de l'organisation prescriptrice de l'émetteur", blank=True, null=True
    )
    sender_prescriber_organization_authorization_status = models.CharField(
        verbose_name="état de l'habilitation de l'organisme prescripteur de l'émetteur", blank=True, null=True
    )

    # company
    company_kind = models.CharField(verbose_name="type d'entreprise")
    company_department = models.CharField(verbose_name="département de l'entreprise")
    company_naf = models.CharField(verbose_name="code NAF de l'entreprise")
    company_has_convention = models.BooleanField(verbose_name="l'entreprise a une convention", default=False)

    # application
    anonymized_at = models.DateTimeField(auto_now_add=True, verbose_name="anonymisé le")
    applied_at = models.DateField(verbose_name="année et mois de la candidature")
    processed_at = models.DateField(verbose_name="année et mois de traitement", blank=True, null=True)
    last_transition_at = models.DateField(
        verbose_name="année et mois de la dernière transition", blank=True, null=True
    )
    had_resume = models.BooleanField(verbose_name="candidature avec un CV", default=False)
    origin = models.CharField(verbose_name="origine de la candidature")
    state = models.CharField(verbose_name="état de la candidature")
    refusal_reason = models.CharField(verbose_name="raison du refus", blank=True, null=True)
    had_been_transferred = models.BooleanField(verbose_name="avait été transférée", default=False)
    number_of_jobs_applied_for = models.PositiveIntegerField(
        verbose_name="nombre d'offres d'emploi pour lesquelles le candidat a postulé", default=0
    )
    had_diagoriente_invitation = models.BooleanField(
        verbose_name="avait reçu une invitation à un diagnostic d'orientation", default=False
    )

    # hiring
    hiring_rome = models.CharField(verbose_name="code ROME de l'embauche", blank=True, null=True)
    hiring_contract_type = models.CharField(verbose_name="type de contrat de l'embauche", blank=True, null=True)
    hiring_contract_nature = models.CharField(verbose_name="nature du contrat de l'embauche", blank=True, null=True)
    hiring_start_date = models.DateField(verbose_name="année et mois de début de l'embauche", blank=True, null=True)

    class Meta:
        verbose_name = "candidature anonymisée"
        verbose_name_plural = "candidatures anonymisées"
        ordering = ["-anonymized_at"]

    def __str__(self):
        return f"candidature {self.id} anonymisée le {self.anonymized_at.strftime('%Y-%m-%d %H:%M:%S')}"
