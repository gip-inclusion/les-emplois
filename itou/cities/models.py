from django.contrib.gis.db import models as gis_models
from django.contrib.postgres.fields import ArrayField
from django.contrib.postgres.indexes import GinIndex
from django.db import models

from itou.common_apps.address.departments import DEPARTMENTS, REGIONS


class EditionModeChoices(models.TextChoices):
    AUTO = "AUTO", "Ville mise à jour automatiquement via script"
    MANUAL = "MANUAL", "Ville mise à jour manuellement"


class City(models.Model):
    """French cities with their geocoding data, synchronized via the sync_cities script regularly."""

    DEPARTMENT_CHOICES = DEPARTMENTS.items()

    name = models.CharField(verbose_name="ville", max_length=255, db_index=True)
    slug = models.SlugField(verbose_name="slug", max_length=255, unique=True)
    department = models.CharField(verbose_name="département", choices=DEPARTMENT_CHOICES, max_length=3, db_index=True)

    # Note that post codes and insee codes have a n-to-n relationship.
    # One insee code can have several post codes but the inverse is also true e.g. zip code 33360 has six insee codes.
    post_codes = ArrayField(models.CharField(max_length=5), verbose_name="codes postaux", blank=True)
    code_insee = models.CharField(verbose_name="code INSEE", max_length=5, unique=True)

    # Latitude and longitude coordinates.
    # https://docs.djangoproject.com/en/2.2/ref/contrib/gis/model-api/#pointfield
    coords = gis_models.PointField(geography=True, blank=True, null=True)

    edition_mode = models.CharField(
        verbose_name="mode d'édition",
        choices=EditionModeChoices.choices,
        max_length=16,
        default=EditionModeChoices.MANUAL,
    )

    objects = models.Manager()  # The default manager.

    class Meta:
        verbose_name = "ville française"
        verbose_name_plural = "villes françaises"
        indexes = [
            # https://docs.djangoproject.com/en/dev/ref/contrib/postgres/search/#trigram-similarity
            # https://docs.djangoproject.com/en/dev/ref/contrib/postgres/indexes/#ginindex
            # https://www.postgresql.org/docs/11/pgtrgm.html#id-1.11.7.40.7
            GinIndex(fields=["name"], name="cities_city_name_gin_trgm", opclasses=["gin_trgm_ops"])
        ]

    def __str__(self):
        return self.display_name

    def save(self, *args, **kwargs):
        # Any save() forces the edition_mode to be MANUAL. Only the sync_cities script
        # can take care of forcing the mode to AUTO.
        self.edition_mode = EditionModeChoices.MANUAL
        super().save(*args, **kwargs)

    @property
    def display_name(self):
        return f"{self.name} ({self.department})"

    def autocomplete_display(self):
        return self.display_name

    @property
    def latitude(self):
        if self.coords:
            return self.coords.y
        return None

    @property
    def longitude(self):
        if self.coords:
            return self.coords.x
        return None

    @property
    def region(self):
        if self.department:
            for region, departments in REGIONS.items():
                if self.department in departments:
                    return region
        return None
