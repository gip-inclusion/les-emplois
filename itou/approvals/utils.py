from itou.companies.enums import SIAE_WITH_CONVENTION_KINDS


def get_user_last_accepted_siae_job_application(user):
    if not user.is_job_seeker:
        return None

    # Some candidates may not have accepted job applications
    # Assuming it's the case can lead to issues downstream
    return (
        user.job_applications.accepted()
        .filter(to_company__kind__in=SIAE_WITH_CONVENTION_KINDS)
        .with_accepted_at()
        .order_by("-accepted_at", "-hiring_start_at")
        .first()
    )


def last_hire_was_made_by_siae(user, siae):
    if not user.is_job_seeker:
        return False
    last_accepted_job_application = get_user_last_accepted_siae_job_application(user)
    return last_accepted_job_application and last_accepted_job_application.to_company_id == siae.pk
