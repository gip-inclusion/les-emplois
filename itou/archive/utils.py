import datetime

from django.db import models
from django.db.models import Count, Exists, OuterRef, Q
from django.db.models.functions import Coalesce
from django.utils import timezone

from itou.approvals.models import Approval
from itou.eligibility.models import EligibilityDiagnosis, GEIQEligibilityDiagnosis
from itou.gps.models import FollowUpGroup
from itou.job_applications.models import JobApplication
from itou.users.models import User, UserKind


def get_filter_kwargs_on_user_for_related_objects_to_check():
    """
    Returns a dictionary of filter parameters to check for objects related
    to the User model that are not cascade deleted.
    """
    return {
        f"{obj.name}__isnull": True
        for obj in User._meta.related_objects
        if getattr(obj, "on_delete", None) and obj.on_delete != models.CASCADE
    }


def get_year_month_or_none(date=None):
    if not date:
        return None

    if isinstance(date, datetime.datetime):
        return timezone.localdate(date).replace(day=1)

    return date.replace(day=1)


def count_related_subquery(model, fk_field, outer_ref_field, extra_filters=None):
    filters = {fk_field: OuterRef(outer_ref_field)}
    if extra_filters:
        filters.update(extra_filters)
    return Coalesce(
        model.objects.filter(**filters).values(fk_field).annotate(count=Count(outer_ref_field)).values("count"), 0
    )


def inactive_jobseekers_without_recent_related_objects(inactive_since, notified, batch_size=None):
    recent_approval = Approval.objects.filter(
        Q(end_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), user_id=OuterRef("pk")
    )
    recent_eligibility_diagnosis = EligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_geiq_eligibility_diagnosis = GEIQEligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_job_application = JobApplication.objects.filter(job_seeker_id=OuterRef("pk"), updated_at__gt=inactive_since)
    recent_followup_group = FollowUpGroup.objects.filter(beneficiary_id=OuterRef("pk"), updated_at__gt=inactive_since)

    qs = (
        User.objects.filter(
            (Q(last_login__lte=inactive_since) | Q(last_login__isnull=True)),
            date_joined__lte=inactive_since,
            kind=UserKind.JOB_SEEKER,
            upcoming_deletion_notified_at__isnull=not notified,
        )
        .filter(
            ~Exists(recent_approval),
            ~Exists(recent_eligibility_diagnosis),
            ~Exists(recent_geiq_eligibility_diagnosis),
            ~Exists(recent_job_application),
            ~Exists(recent_followup_group),
        )
        .order_by("date_joined", "pk")
    )

    if batch_size:
        return qs[:batch_size]

    return qs
