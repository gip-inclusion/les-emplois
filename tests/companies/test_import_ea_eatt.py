import datetime
import io
import re
import zipfile

import pytest

from itou.companies.management.commands import import_ea_eatt
from itou.companies.models import Company
from itou.utils.asp import REMOTE_DOWNLOAD_DIR, REMOTE_UPLOAD_DIR
from itou.utils.date import monday_of_the_week


@pytest.fixture(name="command")
def command_fixture(mocker, settings, sftp_directory, sftp_client_factory):
    # Set required settings
    settings.ASP_SFTP_HOST = "0.0.0.0"  # non-routable IP, just in case :)
    settings.ASP_SFTP_USER = "django_tests"

    # Setup directory
    sftp_directory.joinpath(REMOTE_UPLOAD_DIR).mkdir()
    sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR).mkdir()

    # Create the management command and mock the SFTP connection
    command = import_ea_eatt.Command(stdout=io.StringIO(), stderr=io.StringIO())
    mocker.patch("itou.utils.asp.get_sftp_connection", sftp_client_factory)

    return command


@pytest.fixture(name="archive_file")
def archive_file_fixture():
    archive = io.BytesIO()

    data = [
        "L|ASP|EA|V1.0|20241231|235959|",  # File header
        # Columns headers
        "ITOU|Type d'entreprise adaptée|Numéro de COT/CPOM|Identifiant EA2 Signataire|Siret Signataire"
        "|Dénomination / raison sociale Signataire|Courriel du contact étab signataire|Siret de l'établissement membre"
        "|Dénomination / raison sociale|Identifiant EA2 Etablissement|Numéro entrée ou batiment|Numéro de voie"
        "|Extension de voie|Code voie|Libelle de la voie|Code Postal|Code INSEE commune|",
        # EA in same COT/CPOM than an EATT
        "ITOU|Entreprise Adaptée|00000001|1|00000000000001|Dumas SA|legerdenise@example.org"
        "|00000000000001|Dumas SA|1||1||Rue|DE LA MOTTE|77550|77296|",
        # EATT in same COT/CPOM than an EA
        "ITOU|Entreprise Adaptée Travail Temporaire|00000001|1|00000000000001|Dumas SA|legerdenise@example.org"
        "|00000000000001|Collet SARL|2||1||Rue|DE LA MOTTE|77550|77296|",
        # EAMP - Should be ignored
        "ITOU|Entreprise Adaptée en Milieu Pénitentiaire |00000001|1|00000000000001|Dumas SA|legerdenise@example.org"
        "|00000000000001|Lemaître S.A.S.|3||1||Rue|DE LA MOTTE|77550|77296|",
        # EA duplicate (by SIRET) - Should be ignored
        "ITOU|Entreprise Adaptée|00000001|4|00000000000001|Dumas SA|legerdenise@example.org"
        "|00000000000001|Dumas SA|4||1||Rue|DE LA MOTTE|77550|77296|",
        # File footer
        "Z|ASP|EA|20241231|235959|4|7|",
    ]
    with zipfile.ZipFile(archive, mode="w", compression=zipfile.ZIP_DEFLATED) as fp:
        fp.setpassword(b"password")
        fp.writestr("EA2_ITOU.txt", data="\n".join(data))

    archive.seek(0)
    return archive


def test_retrieve_archive_of_the_week(faker, sftp_directory, command):
    monday = monday_of_the_week()
    monday_archive = sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR, faker.asp_ea2_filename(monday))
    monday_archive.write_bytes(faker.zip())
    assert command.retrieve_archive_of_the_week().getvalue()


def test_retrieve_archive_of_the_week_errors(faker, sftp_directory, command):
    monday = monday_of_the_week()

    with pytest.raises(RuntimeError, match="No file for this week: "):
        command.retrieve_archive_of_the_week()

    sunday_before_monday = monday - datetime.timedelta(days=1)
    sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR, faker.asp_ea2_filename(sunday_before_monday)).touch()
    with pytest.raises(RuntimeError, match="No file for this week: "):
        command.retrieve_archive_of_the_week()

    monday_archive = sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR, faker.asp_ea2_filename(monday))
    monday_archive.write_bytes(faker.zip())
    assert command.retrieve_archive_of_the_week().getvalue()

    tuesday_after_monday = monday + datetime.timedelta(days=1)
    sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR, faker.asp_ea2_filename(tuesday_after_monday)).touch()
    with pytest.raises(RuntimeError, match="Too many files for this week: "):
        command.retrieve_archive_of_the_week()


def test_clean_old_archives(faker, mocker, sftp_directory, command):
    expected_files = set()
    for day in range(3):
        filename = faker.asp_ea2_filename(datetime.date(2024, 8, 1 + day))
        sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR, filename).touch()
        expected_files.add(filename)

    assert {file.name for file in sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR).iterdir()} == expected_files
    command.clean_old_archives(wet_run=True)
    assert {file.name for file in sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR).iterdir()} == expected_files

    mocker.patch.object(command, "NUMBER_OF_ARCHIVES_TO_KEEP", 1)
    command.clean_old_archives(wet_run=True)
    assert {file.name for file in sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR).iterdir()} == {
        faker.asp_ea2_filename(datetime.date(2024, 8, 3))
    }


def test_clean_old_archives_dry_run(faker, mocker, sftp_directory, command):
    mocker.patch.object(command, "NUMBER_OF_ARCHIVES_TO_KEEP", 0)
    sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR, faker.asp_ea2_filename(monday_of_the_week())).touch()

    assert len(set(sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR).iterdir())) == 1
    command.clean_old_archives(wet_run=False)
    assert len(set(sftp_directory.joinpath(REMOTE_DOWNLOAD_DIR).iterdir())) == 1


def test_process_file_from_archive(capsys, snapshot, settings, command, archive_file):
    settings.ASP_EA2_UNZIP_PASSWORD = "password"

    command.handle(from_archive=archive_file, wet_run=True)
    assert re.sub(r"siae.id=\d+", "siae.id=[ID]", capsys.readouterr()[0] + command.stdout.getvalue()) == snapshot(
        name="output"
    )
    filled_fields = [
        "kind",
        "siret",
        "source",
        "name",
        "email",
        "auth_email",
        "address_line_1",
        "address_line_2",
        "post_code",
        "insee_city",
        "city",
        "department",
        "coords",
    ]
    assert list(Company.objects.all().order_by("pk").values_list(*filled_fields)) == snapshot(name="data")
