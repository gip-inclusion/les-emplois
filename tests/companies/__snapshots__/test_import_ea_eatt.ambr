# serializer version: 1
# name: test_process_file_from_archive[data]
  list([
    tuple(
      'EA',
      '00000000000001',
      'EA_EATT',
      'Dumas Sa',
      '',
      'legerdenise@example.org',
      '1 Rue DE LA MOTTE',
      '',
      '77550',
      None,
      '',
      '77',
      None,
    ),
    tuple(
      'EATT',
      '00000000000001',
      'EA_EATT',
      'Collet Sarl',
      '',
      'legerdenise@example.org',
      '1 Rue DE LA MOTTE',
      '',
      '77550',
      None,
      '',
      '77',
      None,
    ),
  ])
# ---
# name: test_process_file_from_archive[output]
  '''
  Loaded 2 EA_EATT from export.
  1 EA_EATT will be created.
  0 EA_EATT will be updated when needed.
  0 EA_EATT will be deleted when possible.
  EA_EATT siret=00000000000001 will be created.
  EA_EATT siret=00000000000001 has been created with siae.id=[ID].
  EA_EATT siret=00000000000001 will be created.
  EA_EATT siret=00000000000001 has been created with siae.id=[ID].
  0 EA_EATT can and will be deleted.
  0 EA_EATT cannot be deleted as they have data.
  --------------------------------------------------------------------------------
  Rows in file: 4
  Rows after kind filter: 3
  Rows after deduplication: 2
  Rows used: 2
   > With a SIRET: 2
   > With an empty email: 0
   > Creatable: 1
   >> Created: 2
   > Updatable: 0
   >> Updated: 0
   > Deletable: 0
   >> Deleted: 0
   >> Undeletable: 0
   >> Skipped: 0
  --------------------------------------------------------------------------------
  Done.
  
  '''
# ---
